### DISCUSSION TITLE
the format of a multi-GTP session

### DISCUSSION THREAD

**Discussion by HackYardo** [kaorahi/lizgoban#80] [Q&A] [answered]
I bring **dot syntax** to multiple GoTextProtocol sessions like this:
```
...
GTP ready
?
=
status, st  : list all bot
  open, run : run a bot
switch, turn: switch to a bot
  help, ?   : show this message
gtp_command.: broadcast a GTP command to all running bots
-----Example:
name.
g = GNU Go
l = Leela Zero

st
= ka g k ka2 l k2 ks

name.
g = GNU Go
k = KataGo
l = Leela Zero

quit.
k =
g =
l =

```

May I ask your advice about it?

**Comment by kaorahi** [comment, marked as answer]
Hmm, I'm not sure what is the merit of it. Existing GUIs (Lizzie, Sabaki, etc.) can switch engines without such GTP extensions.

Anyway, this seems off-topic for lizgoban and I cannot promise my further response...

**Reply by HackYardo** [reply]
These two are different: one can only ask one backend at a time, and the other can ask all running backends at a time, so that the data reported by multiple backends are presented to the frontend at the same time.

You could say I made this is for providing a tool for the community, but initially it was something I want to use myself. I'm also not here to sell anything, but to consult the community's predecessors on what format would be easy for both machine parsing and human reading.

Anyway, thanks for your reply.

**Comment by kaorahi** [comment]
Please understand I need some efforts to write English. So I cannot promise my reply to every comment.

I'm still not sure what is the merit of (2) compared with (1). (2) requires some programming cost for GUI to support the extended GTP. It seems as expensive as that of multiple GTP sessions in (1) for me.

![extendedGTP](https://user-images.githubusercontent.com/38910552/200540711-83501231-7ce4-4d30-9dc8-c47a91c893fc.jpg)

If you are interested in development of the component `[???]`, (3) may be more worthy so that GUI developers don't need to parse GTP by themselves. In lizgoban, `ai.js` corresponds to `[???]` in (3). It offers several convenient features whose implementations are annoying for Go GUIs.

1. parse GTP
2. emulate a stateless protocol like the [analysis engine](https://github.com/lightvector/KataGo/blob/v1.11.0/docs/Analysis_Engine.md) of katago
3. keep several engines in background so that they can be switched quickly
4. restart the engine automatically when it has a trouble

In particular, 2 is very useful to avoid synchronization bugs between the boards of GUI and the engine. I repeatedly experienced such bugs in lizzie.

`ai.js` is not separated well from other parts of lizgoban unfortunately. If you write an independent `[???]` in (3) that can be used from any language, it may help future programmers for easy development of their own Go GUIs. (though I'm not sure...)

**Comment by HackYardo** [comment]
- What is your first language? If you don't familiar with English, you can try translate tool such as https://cn.bing.com/translator, https://www.deepl.com/translator and https://translate.yandex.com.

- Thanks for your reply and sorry for my late reply. 

- I don't get the point, what's the difference between (2) and (3)? By JSON format, you mean a dict of key-value pairs, which is easy to parse and print than raw GTP inputs/outputs? `JSON IO > GTP IO` ?  Good idea, but it is easy for computer while hard for people from my view, e.g. JSON has no matrix and D>2 array format, so people can not see the board status properly like the GTP `showboard` does.

**Reply by kaorahi** [reply]
You seem to assume direct use of GTP by hand without GUI. Then I have no opinion as I seldom use GTP in such a way.
